relay-server/src/utils/rate_limits.rs

```diff
@@ -816,18 +901,115 @@ mod tests {
+    fn test_transaction_indexed_only() {
+        let mut envelope = envelope![Transaction];
+        let mut mock = MockLimiter::default().deny(DataCategory::Transaction);
+        let limiter = EnvelopeLimiter::new(|s, q| mock.check(s, q));
+
+        let (enforcement, limits) = limiter
+            .enforce(&mut envelope, &scoping(), &Enforcement::default())
+            .unwrap();
+
+        assert!(enforcement.event.is_active());
+        assert!(!enforcement.transaction_processed.is_active());
+        assert!(!limits.is_limited());
+        assert!(!envelope.is_empty());
+        mock.assert_call(DataCategory::Transaction, Some(1));
+        mock.assert_call(DataCategory::TransactionProcessed, Some(1));
+    }
+
+    #[test]
+    #[cfg(feature = "processing")]
+    fn test_transaction_processed_only() {
+        let mut envelope = envelope![Transaction];
+        let mut mock = MockLimiter::default().deny(DataCategory::TransactionProcessed);
+        let limiter = EnvelopeLimiter::new(|s, q| mock.check(s, q));
+
+        let (enforcement, limits) = limiter
+            .enforce(&mut envelope, &scoping(), &Enforcement::default())
+            .unwrap();
+
+        assert!(!enforcement.event.is_active());
+        assert!(enforcement.transaction_processed.is_active());
+        assert!(!limits.is_limited());
+        assert!(!envelope.is_empty());
+        mock.assert_call(DataCategory::Transaction, Some(1));
+        mock.assert_call(DataCategory::TransactionProcessed, Some(1));
+    }
+
+    #[test]
+    #[cfg(feature = "processing")]
+    fn test_transaction_early_enforcement() {
+        let mut envelope = envelope![Transaction];
+
+        // Early enforcement
+        let mut mock = MockLimiter::default().deny(DataCategory::Transaction);
+        let limiter = EnvelopeLimiter::new(|s, q| mock.check(s, q));
+
+        let (enforcement, limits) = limiter
+            .enforce(&mut envelope, &scoping(), &Enforcement::default())
+            .unwrap();
+
+        assert!(enforcement.event.is_active());
+        assert!(!enforcement.transaction_processed.is_active());
+        assert!(!limits.is_limited());
+        assert!(!envelope.is_empty());
+        mock.assert_call(DataCategory::Transaction, Some(1));
+        mock.assert_call(DataCategory::TransactionProcessed, Some(1));
+
+        // 2nd quota check
+        let mut mock = MockLimiter::default().deny(DataCategory::Transaction);
+        let limiter = EnvelopeLimiter::new(|s, q| mock.check(s, q));
+
+        let (enforcement, limits) = limiter
+            .enforce(&mut envelope, &scoping(), &enforcement)
+            .unwrap();
+
+        assert!(enforcement.event.is_active());
+        assert!(!enforcement.transaction_processed.is_active());
+        assert!(!limits.is_limited());
+        assert!(!envelope.is_empty());
+        mock.assert_not_called(DataCategory::Transaction);
+        mock.assert_call(DataCategory::TransactionProcessed, Some(1));
+    }
```

Comment:
General question on tests: Do we need an integration test for any of this, or do we feel confident that the unit tests cover everything?